### PR TITLE
fix use-after-free in normalizeNoteSegments

### PR DESCRIPTION
**Pull request by Mic92**

### CODE REVIEW

**Comment by andir** [comment]
I can see how iterating while mutating is bad but care to explain the use after free?

**Comment by Mic92** [comment]
You can spot the bug by enabling `./configure --prefix=$(pwd)/inst --with-asan` and than do a `make check`

```
+ ARCH=
+ PAGESIZE=4096
+ '[' -z '' ']'
++ basename ./no-rpath-amd64.sh .sh
++ sed -e 's/^no-rpath-//'
+ ARCH=amd64
+ SCRATCH=scratch/no-rpath-amd64
+ '[' -z amd64 ']'
+ '[' amd64 = prebuild ']'
+ no_rpath_bin=./no-rpath-prebuild/no-rpath-amd64
+ '[' '!' -f ./no-rpath-prebuild/no-rpath-amd64 ']'
+ rm -rf scratch/no-rpath-amd64
+ mkdir -p scratch/no-rpath-amd64
+ cp ./no-rpath-prebuild/no-rpath-amd64 scratch/no-rpath-amd64/no-rpath
++ ../src/patchelf --page-size 4096 --print-rpath scratch/no-rpath-amd64/no-rpath
+ oldRPath=
+ test -n ''
++ ../src/patchelf --page-size 4096 --print-interpreter ../src/patchelf
+ ../src/patchelf --page-size 4096 --set-interpreter /nix/store/gk42f59363p82rg2wv2mfy71jn5w4q4c-glibc-2.32-48/lib/ld-linux-x86-64.so.2 --set-rpath /foo:/bar:/xxxxxxxxxxxxxxx scratch/no-rpath-amd64/no-rpath
patching ELF file 'scratch/no-rpath-amd64/no-rpath'
new rpath is '/foo:/bar:/xxxxxxxxxxxxxxx'
rpath is too long, resizing...
DT_NULL index is 24
replacing section '.dynamic' with size 496
replacing section '.dynstr' with size 88
replacing section '.interp' with size 83
this is an executable
using replaced section '.interp'
using replaced section '.dynstr'
using replaced section '.dynamic'
last replaced is 22
looking at section '.interp'
looking at section '.note.ABI-tag'
replacing section '.note.ABI-tag' which is in the way
looking at section '.note.gnu.build-id'
replacing section '.note.gnu.build-id' which is in the way
looking at section '.hash'
replacing section '.hash' which is in the way
looking at section '.gnu.hash'
replacing section '.gnu.hash' which is in the way
looking at section '.dynsym'
replacing section '.dynsym' which is in the way
looking at section '.dynstr'
looking at section '.gnu.version'
first reserved offset/addr is 0x346/0x400346
first page is 0x400000
=================================================================
==3854456==ERROR: AddressSanitizer: heap-use-after-free on address 0x614000000190 at pc 0x000000447f94 bp 0x7ffd8bf1be60 sp 0x7ffd8bf1be58
READ of size 4 at 0x614000000190 thread T0
    #0 0x447f93 in ElfFile<Elf64_Ehdr, Elf64_Phdr, Elf64_Shdr, unsigned long, unsigned long, Elf64_Dyn, Elf64_Sym, Elf64_Verneed, unsigned short>::normalizeNoteSegments() /home/joerg/git/patchelf/src/patchelf.cc:1029
    #1 0x47344a in ElfFile<Elf64_Ehdr, Elf64_Phdr, Elf64_Shdr, unsigned long, unsigned long, Elf64_Dyn, Elf64_Sym, Elf64_Verneed, unsigned short>::rewriteSectionsExecutable() /home/joerg/git/patchelf/src/patchelf.cc:969
    #2 0x474475 in ElfFile<Elf64_Ehdr, Elf64_Phdr, Elf64_Shdr, unsigned long, unsigned long, Elf64_Dyn, Elf64_Sym, Elf64_Verneed, unsigned short>::rewriteSections() /home/joerg/git/patchelf/src/patchelf.cc:1087
    #3 0x4120ce in patchElf2<ElfFile<Elf64_Ehdr, Elf64_Phdr, Elf64_Shdr, long unsigned int, long unsigned int, Elf64_Dyn, Elf64_Sym, Elf64_Verneed, short unsigned int> > /home/joerg/git/patchelf/src/patchelf.cc:1840
    #4 0x4120ce in patchElf /home/joerg/git/patchelf/src/patchelf.cc:1861
    #5 0x4120ce in mainWrapped(int, char**) /home/joerg/git/patchelf/src/patchelf.cc:2016
    #6 0x40542e in main /home/joerg/git/patchelf/src/patchelf.cc:2024
    #7 0x7fefd9d9edec in __libc_start_main (/nix/store/gk42f59363p82rg2wv2mfy71jn5w4q4c-glibc-2.32-48/lib/libc.so.6+0x27dec)
    #8 0x405df9 in _start (/home/joerg/git/patchelf/src/patchelf+0x405df9)

0x614000000190 is located 336 bytes inside of 448-byte region [0x614000000040,0x614000000200)
freed by thread T0 here:
    #0 0x7fefda318fd7 in operator delete(void*) (/nix/store/dzaadhn5y56da24635icafv1nrawxm4n-gcc-10.3.0-lib/lib/libasan.so.6+0xaefd7)
    #1 0x41f51d in __gnu_cxx::new_allocator<Elf64_Phdr>::deallocate(Elf64_Phdr*, unsigned long) /nix/store/5qjycalzb9sqzvqg65kf5zimqwjabm9g-gcc-10.3.0/include/c++/10.3.0/ext/new_allocator.h:133
    #2 0x41f51d in std::allocator_traits<std::allocator<Elf64_Phdr> >::deallocate(std::allocator<Elf64_Phdr>&, Elf64_Phdr*, unsigned long) /nix/store/5qjycalzb9sqzvqg65kf5zimqwjabm9g-gcc-10.3.0/include/c++/10.3.0/bits/alloc_traits.h:492
    #3 0x41f51d in std::_Vector_base<Elf64_Phdr, std::allocator<Elf64_Phdr> >::_M_deallocate(Elf64_Phdr*, unsigned long) /nix/store/5qjycalzb9sqzvqg65kf5zimqwjabm9g-gcc-10.3.0/include/c++/10.3.0/bits/stl_vector.h:354
    #4 0x41f51d in void std::vector<Elf64_Phdr, std::allocator<Elf64_Phdr> >::_M_realloc_insert<Elf64_Phdr const&>(__gnu_cxx::__normal_iterator<Elf64_Phdr*, std::vector<Elf64_Phdr, std::allocator<Elf64_Phdr> > >, Elf64_Phdr const&) /nix/store/5qjycalzb9sqzvqg65kf5zimqwjabm9g-gcc-10.3.0/include/c++/10.3.0/bits/vector.tcc:500
    #5 0x447d76 in std::vector<Elf64_Phdr, std::allocator<Elf64_Phdr> >::push_back(Elf64_Phdr const&) /nix/store/5qjycalzb9sqzvqg65kf5zimqwjabm9g-gcc-10.3.0/include/c++/10.3.0/bits/stl_vector.h:1198
    #6 0x447d76 in std::vector<Elf64_Phdr, std::allocator<Elf64_Phdr> >::push_back(Elf64_Phdr const&) /nix/store/5qjycalzb9sqzvqg65kf5zimqwjabm9g-gcc-10.3.0/include/c++/10.3.0/bits/stl_vector.h:1187
    #7 0x447d76 in ElfFile<Elf64_Ehdr, Elf64_Phdr, Elf64_Shdr, unsigned long, unsigned long, Elf64_Dyn, Elf64_Sym, Elf64_Verneed, unsigned short>::normalizeNoteSegments() /home/joerg/git/patchelf/src/patchelf.cc:1064
    #8 0x47344a in ElfFile<Elf64_Ehdr, Elf64_Phdr, Elf64_Shdr, unsigned long, unsigned long, Elf64_Dyn, Elf64_Sym, Elf64_Verneed, unsigned short>::rewriteSectionsExecutable() /home/joerg/git/patchelf/src/patchelf.cc:969
    #9 0x474475 in ElfFile<Elf64_Ehdr, Elf64_Phdr, Elf64_Shdr, unsigned long, unsigned long, Elf64_Dyn, Elf64_Sym, Elf64_Verneed, unsigned short>::rewriteSections() /home/joerg/git/patchelf/src/patchelf.cc:1087
    #10 0x4120ce in patchElf2<ElfFile<Elf64_Ehdr, Elf64_Phdr, Elf64_Shdr, long unsigned int, long unsigned int, Elf64_Dyn, Elf64_Sym, Elf64_Verneed, short unsigned int> > /home/joerg/git/patchelf/src/patchelf.cc:1840
    #11 0x4120ce in patchElf /home/joerg/git/patchelf/src/patchelf.cc:1861
    #12 0x4120ce in mainWrapped(int, char**) /home/joerg/git/patchelf/src/patchelf.cc:2016
    #13 0x40542e in main /home/joerg/git/patchelf/src/patchelf.cc:2024
    #14 0x7fefd9d9edec in __libc_start_main (/nix/store/gk42f59363p82rg2wv2mfy71jn5w4q4c-glibc-2.32-48/lib/libc.so.6+0x27dec)

previously allocated by thread T0 here:
    #0 0x7fefda318607 in operator new(unsigned long) (/nix/store/dzaadhn5y56da24635icafv1nrawxm4n-gcc-10.3.0-lib/lib/libasan.so.6+0xae607)
    #1 0x41f2bc in __gnu_cxx::new_allocator<Elf64_Phdr>::allocate(unsigned long, void const*) /nix/store/5qjycalzb9sqzvqg65kf5zimqwjabm9g-gcc-10.3.0/include/c++/10.3.0/ext/new_allocator.h:115
    #2 0x41f2bc in std::allocator_traits<std::allocator<Elf64_Phdr> >::allocate(std::allocator<Elf64_Phdr>&, unsigned long) /nix/store/5qjycalzb9sqzvqg65kf5zimqwjabm9g-gcc-10.3.0/include/c++/10.3.0/bits/alloc_traits.h:460
    #3 0x41f2bc in std::_Vector_base<Elf64_Phdr, std::allocator<Elf64_Phdr> >::_M_allocate(unsigned long) /nix/store/5qjycalzb9sqzvqg65kf5zimqwjabm9g-gcc-10.3.0/include/c++/10.3.0/bits/stl_vector.h:346
    #4 0x41f2bc in void std::vector<Elf64_Phdr, std::allocator<Elf64_Phdr> >::_M_realloc_insert<Elf64_Phdr const&>(__gnu_cxx::__normal_iterator<Elf64_Phdr*, std::vector<Elf64_Phdr, std::allocator<Elf64_Phdr> > >, Elf64_Phdr const&) /nix/store/5qjycalzb9sqzvqg65kf5zimqwjabm9g-gcc-10.3.0/include/c++/10.3.0/bits/vector.tcc:440
    #5 0x420b0f in std::vector<Elf64_Phdr, std::allocator<Elf64_Phdr> >::push_back(Elf64_Phdr const&) /nix/store/5qjycalzb9sqzvqg65kf5zimqwjabm9g-gcc-10.3.0/include/c++/10.3.0/bits/stl_vector.h:1198
    #6 0x420b0f in std::vector<Elf64_Phdr, std::allocator<Elf64_Phdr> >::push_back(Elf64_Phdr const&) /nix/store/5qjycalzb9sqzvqg65kf5zimqwjabm9g-gcc-10.3.0/include/c++/10.3.0/bits/stl_vector.h:1187
    #7 0x420b0f in ElfFile<Elf64_Ehdr, Elf64_Phdr, Elf64_Shdr, unsigned long, unsigned long, Elf64_Dyn, Elf64_Sym, Elf64_Verneed, unsigned short>::ElfFile(std::shared_ptr<std::vector<unsigned char, std::allocator<unsigned char> > >) /home/joerg/git/patchelf/src/patchelf.cc:420
    #8 0x411095 in patchElf /home/joerg/git/patchelf/src/patchelf.cc:1861
    #9 0x411095 in mainWrapped(int, char**) /home/joerg/git/patchelf/src/patchelf.cc:2016
    #10 0x40542e in main /home/joerg/git/patchelf/src/patchelf.cc:2024
    #11 0x7fefd9d9edec in __libc_start_main (/nix/store/gk42f59363p82rg2wv2mfy71jn5w4q4c-glibc-2.32-48/lib/libc.so.6+0x27dec)

SUMMARY: AddressSanitizer: heap-use-after-free /home/joerg/git/patchelf/src/patchelf.cc:1029 in ElfFile<Elf64_Ehdr, Elf64_Phdr, Elf64_Shdr, unsigned long, unsigned long, Elf64_Dyn, Elf64_Sym, Elf64_Verneed, unsigned short>::normalizeNoteSegments()
Shadow bytes around the buggy address:
  0x0c287fff7fe0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x0c287fff7ff0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x0c287fff8000: fa fa fa fa fa fa fa fa fd fd fd fd fd fd fd fd
  0x0c287fff8010: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
  0x0c287fff8020: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
=>0x0c287fff8030: fd fd[fd]fd fd fd fd fd fd fd fd fd fd fd fd fd
  0x0c287fff8040: fa fa fa fa fa fa fa fa fd fd fd fd fd fd fd fd
  0x0c287fff8050: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
  0x0c287fff8060: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
  0x0c287fff8070: fd fd fd fd fd fd fd fd fd fd fd fa fa fa fa fa
  0x0c287fff8080: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
Shadow byte legend (one shadow byte represents 8 application bytes):
  Addressable:           00
  Partially addressable: 01 02 03 04 05 06 07 
  Heap left redzone:       fa
  Freed heap region:       fd
  Stack left redzone:      f1
  Stack mid redzone:       f2
  Stack right redzone:     f3
  Stack after return:      f5
  Stack use after scope:   f8
  Global redzone:          f9
  Global init order:       f6
  Poisoned by user:        f7
  Container overflow:      fc
  Array cookie:            ac
  Intra object redzone:    bb
  ASan internal:           fe
  Left alloca redzone:     ca
  Right alloca redzone:    cb
  Shadow gap:              cc
==3854456==ABORTING
FAIL no-rpath-amd64.sh (exit status: 1)
```

File: src/patchelf.cc [hunk 1]
```diff
@@ -1011,6 +1011,7 @@ void ElfFile<ElfFileParamNames>::normalizeNoteSegments()
         [this](std::pair<const std::string, std::string> & i) { return rdi(findSection(i.first).sh_type) == SHT_NOTE; });
     if (!replaced_note) return;
 
+    std::vector<Elf_Phdr> newPhdrs;
     for (auto & phdr : phdrs) {
         if (rdi(phdr.p_type) != PT_NOTE) continue;
```
Comment (Mic92) [inline src/patchelf.cc:1016]: The use after free occurs here

File: src/patchelf.cc [hunk 2]
```diff
@@ -1047,11 +1048,13 @@ void ElfFile<ElfFileParamNames>::normalizeNoteSegments()
             if (curr_off == start_off)
                 phdr = new_phdr;
             else
-                phdrs.push_back(new_phdr);
```
Comment (Mic92) [inline src/patchelf.cc:1050]: And the free happend here

**Comment by Mic92** [comment]
This behavior is called `Iterator invalidation` in C++.

**Pull request merged**